File: 349-intersection-of-two-arrays/349-intersection-of-two-arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */

// naive approach
// var intersection = function(nums1, nums2) {
//     let result = [];
//     for (let i = 0; i < nums1.length; i++){
//         if (result.includes(nums1[i])) continue;
//         if (nums2.includes(nums1[i])){
//             result.push(nums1[i]);
//         } else {
//             continue;
//         }
//     }
//     return result;
// };

// use of sets
var intersection = function(nums1, nums2) {
    const compareSets = function(smallSet, largeSet){
        let result = [];
        for (let i = 0; i < smallSet.length; i++){
            if (largeSet.includes(smallSet[i])) result.push(smallSet[i]);
        }
        return result;
    }
    let set1 = [...new Set(nums1)]
    let set2 = [...new Set(nums2)]
    
    if (set1.length > set2.length){
        return compareSets(set2, set1);
    } else {
        return compareSets(set1, set2);
    }
};

File: 349-intersection-of-two-arrays/349-intersection-of-two-arrays.py
class Solution:
    def compare_sets(self, small_set: Set[int], large_set: Set[int]) -> List[int]:
        result = []
        for i in small_set:
            if i in large_set:
                result.append(i)
        return result
    
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        set1 = set(nums1)
        set2 = set(nums2)
        if len(set1) > len(set2):
            return self.compare_sets(set2, set1)
        else:
            return self.compare_sets(set1, set2)